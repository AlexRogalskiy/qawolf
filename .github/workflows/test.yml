name: Test
on: push
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v1
        with:
          ref: ${{ github.ref }}
      # build first to partition the action logs
      - name: Build
        uses: ./
        with:
          args: 'echo "Dockerfile built"'
      - name: Test
        uses: ./
        with:
          args: "cd $QAWOLF_DIR && npm test"
        env:
          LOG_PATH: /github/workspace/artifacts/logs
          SLEEP_AFTER_EACH: 0
          TEST_URL: http://theinternet:5000/
      - name: Upload Artifacts
        if: always()
        uses: actions/upload-artifact@master
        with:
          name: logs
          path: "./artifacts/logs"
    services:
      theinternet:
        image: gprestes/the-internet@sha256:f741ba048b7232ef90b0f7885de7f6ac42a52a17e73dd82afc484da140398796
