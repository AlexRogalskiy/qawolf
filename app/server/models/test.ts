import isNil from "lodash/isNil";

import { minutesFromNow } from "../../shared/utils";
import { db } from "../db";
import { ClientError } from "../errors";
import { ModelOptions, Test } from "../types";
import { cuid } from "../utils";

type BuildTestName = {
  name?: string;
  team_id: string;
};

type CreateTest = {
  code: string;
  creator_id: string;
  group_ids: string[];
  name?: string;
  team_id: string;
  url: string;
};

type FindEnabledTestsForGroup = {
  group_id: string;
  test_ids?: string[] | null;
};

export type LocationCount = {
  count: number;
  location: string | null;
};

type UpdateTest = {
  code?: string;
  id: string;
  is_enabled?: boolean;
  name?: string;
  runner_requested_at?: null;
  version?: number;
};

type UpdateTestToPending = {
  id: string;
  runner_locations: string[];
};

export const buildTestName = async (
  { name, team_id }: BuildTestName,
  { logger, trx }: ModelOptions
): Promise<string> => {
  const tests = await findTestsForTeam(team_id, { logger, trx });

  const testNames = new Set(tests.map((test) => test.name));
  // use current name if possible
  if (name && !testNames.has(name)) return name;

  let testNumber = tests.length + 1;

  while (testNames.has(`My Test${testNumber === 1 ? "" : ` ${testNumber}`}`)) {
    testNumber++;
  }

  return `My Test${testNumber === 1 ? "" : ` ${testNumber}`}`;
};

export const countPendingTests = async (
  defaultLocation: string,
  { logger, trx }: ModelOptions
): Promise<LocationCount[]> => {
  const log = logger.prefix("countPendingTests");

  const countQuery = `SELECT COUNT(s.*), s.location
  FROM (
    SELECT coalesce(runner_locations::json->> 0, ?) as location
    FROM tests
    WHERE runner_requested_at IS NOT NULL 
      AND deleted_at IS NULL
  ) as s
  GROUP BY s.location`;

  const { rows } = await (trx || db).raw(countQuery, [defaultLocation]);

  const result = rows.map((row) => ({
    count: Number(row.count),
    location: row.location,
  }));

  log.debug(result);

  return result;
};

export const createTestAndGroupTests = async (
  { code, creator_id, group_ids, name, team_id, url }: CreateTest,
  { logger, trx }: ModelOptions
): Promise<{ groupTestIds: string[]; test: Test }> => {
  const log = logger.prefix("createTestAndGroupTest");

  log.debug("creator", creator_id);
  const timestamp = minutesFromNow();

  const { groupTestIds, test } = await (trx || db).transaction(async (trx) => {
    const finalName = await buildTestName({ name, team_id }, { logger, trx });

    const test: Test = {
      created_at: timestamp,
      creator_id,
      code,
      deleted_at: null,
      id: cuid(),
      is_enabled: true,
      name: finalName,
      team_id,
      url,
      updated_at: timestamp,
      version: 0,
    };

    await trx("tests").insert(test);

    const groupTests = group_ids.map((group_id) => {
      return { group_id, id: cuid(), test_id: test.id };
    });
    await trx("group_tests").insert(groupTests);

    return { groupTestIds: groupTests.map((g) => g.id), test };
  });

  log.debug("created test", test.id);

  return { groupTestIds, test };
};

export const findTestsForGroup = async (
  group_id: string,
  { logger, trx }: ModelOptions
): Promise<Test[]> => {
  const log = logger.prefix("findTestsForGroup");

  log.debug(group_id);

  const tests = await (trx || db)
    .select("tests.*" as "*")
    .from("tests")
    .innerJoin("group_tests", "group_tests.test_id", "tests.id")
    .where({ deleted_at: null, "group_tests.group_id": group_id })
    .orderBy("name", "asc");

  log.debug(`found ${tests.length} tests for group ${group_id}`);

  return tests;
};

export const findTestsForTeam = async (
  team_id: string,
  { logger, trx }: ModelOptions
): Promise<Test[]> => {
  const log = logger.prefix("findTestsForTeam");

  log.debug(team_id);

  const tests = await (trx || db)
    .select("*")
    .from("tests")
    .where({ deleted_at: null, team_id })
    .orderBy("name", "asc");

  log.debug(`found ${tests.length} tests for team ${team_id}`);

  return tests;
};

export const findEnabledTestsForGroup = async (
  { group_id, test_ids }: FindEnabledTestsForGroup,
  { logger, trx }: ModelOptions
): Promise<Test[]> => {
  const log = logger.prefix("findEnabledTestsForGroup");
  log.debug(group_id);

  const query = (trx || db)
    .select("tests.*" as "*")
    .from("tests")
    .innerJoin("group_tests", "group_tests.test_id", "tests.id")
    .where({ deleted_at: null, group_id, is_enabled: true });

  if (test_ids && test_ids.length) {
    query.whereIn("tests.id", test_ids);
  }

  const tests = await query.orderBy("created_at", "asc");

  log.debug(`found ${tests.length} enabled tests for group ${group_id}`);

  return tests;
};

export const findPendingTest = async (
  location: string | null,
  { logger, trx }: ModelOptions
): Promise<Pick<Test, "id" | "runner_requested_at"> | null> => {
  const log = logger.prefix("findPendingTest");

  let query = (trx || db)("tests")
    .select("id")
    .select("runner_requested_at")
    .whereNotNull("runner_requested_at")
    .where({ deleted_at: null });

  if (location) {
    query = query.where("runner_locations", "like", `%${location}%`);
  }

  const result = await query.orderBy("runner_requested_at", "asc").first();

  if (result) log.debug("found", result, "for location", location);

  return result || null;
};

export const findTest = async (
  id: string,
  { logger, trx }: ModelOptions
): Promise<Test> => {
  const log = logger.prefix("findTest");

  if (!id) throw new Error("Id required");

  const test = await (trx || db)
    .select("*")
    .from("tests")
    .where({ id })
    .first();

  if (!test) {
    log.error("not found", id);
    throw new Error(`test not found ${id}`);
  }

  if (test.runner_locations) {
    test.runner_locations = JSON.parse(test.runner_locations);
  }

  return test;
};

export const findTestForRun = async (
  run_id: string,
  { logger, trx }: ModelOptions
): Promise<Test> => {
  const log = logger.prefix("findTestForRun");

  const test = await (trx || db)
    .select("tests.*" as "*")
    .from("tests")
    .innerJoin("runs", "runs.test_id", "tests.id")
    .where({ "runs.id": run_id })
    .first();

  if (!test) {
    log.error("not found", run_id);
    throw new Error(`test for run not found ${run_id}`);
  }

  return test;
};

export const deleteTests = async (
  ids: string[],
  { logger, trx }: ModelOptions
): Promise<Test[]> => {
  const log = logger.prefix("deleteTests");

  log.debug(ids);

  const tests = await (trx || db).transaction(async (trx) => {
    const tests = await trx.select("*").from("tests").whereIn("id", ids);
    const updates = { deleted_at: minutesFromNow() };
    await trx("tests").update(updates).whereIn("id", ids);
    return tests.map((test: Test) => ({ ...test, ...updates }));
  });

  log.debug("deleted", ids);

  return tests;
};

export const updateTest = async (
  { code, id, is_enabled, name, runner_requested_at, version }: UpdateTest,
  { logger, trx }: ModelOptions
): Promise<Test> => {
  const log = logger.prefix("updateTest");

  return (trx || db).transaction(async (trx) => {
    const existingTest = await findTest(id, { logger, trx });

    if (!existingTest) {
      log.debug("test not found", id);
      throw new Error(`Test not found ${id}`);
    }

    // do not overwrite current version with older version
    /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */
    if (!isNil(version) && existingTest.version >= version!) {
      log.debug(
        `ignore: test ${id} current version ${existingTest.version} >= update version ${version}`
      );
      return existingTest;
    }

    const updates: Partial<Test> = {
      code: code || existingTest.code,
      is_enabled: isNil(is_enabled) ? existingTest.is_enabled : is_enabled,
      name: name || existingTest.name,
      updated_at: minutesFromNow(),
      version: version || existingTest.version,
    };

    if (runner_requested_at === null) {
      updates.runner_requested_at = runner_requested_at;
    }

    try {
      await trx("tests").where({ id }).update(updates);
    } catch (error) {
      if (error.message.includes("tests_unique_name_team_id")) {
        throw new ClientError("test name must be unique");
      }

      throw error;
    }

    log.debug("updated", id, updates);

    return { ...existingTest, ...updates };
  });
};

export const updateTestToPending = async (
  options: UpdateTestToPending,
  { logger, trx }: ModelOptions
): Promise<boolean> => {
  const log = logger.prefix("updateTestToPending");

  const runner_locations = options.runner_locations?.length
    ? JSON.stringify(options.runner_locations.slice(0, 2))
    : null;

  // do not update if there is a runner assigned
  // this could happen during a concurrent assignment
  const sql = `UPDATE tests
  SET runner_locations = ?, runner_requested_at = now()
  WHERE id = ? 
    AND runner_requested_at IS NULL 
    AND id NOT IN (SELECT test_id FROM runners)`;

  const result = await (trx || db).raw(sql, [runner_locations, options.id]);

  const didUpdate = result.rowCount > 0;
  log.debug(options, didUpdate ? "updated" : "skipped");

  return didUpdate;
};
