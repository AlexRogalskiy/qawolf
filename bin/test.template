const { Browser, hooks } = require("{{modulePath}}");

let browser;

beforeAll(async () => {
  browser = await Browser.create({ size: "{{{size}}}", url: "{{{url}}}" });
  await hooks.beforeAll(browser);
});

beforeEach(() => hooks.beforeEach(browser));

afterEach(() => hooks.afterEach(browser));

afterAll(() => hooks.afterAll(browser));

{{#each formattedSteps}}
test('{{{name}}}', async () => {
  await browser.runStep({{{json}}});
});
{{#unless @last}}

{{/unless}}
{{/each}}