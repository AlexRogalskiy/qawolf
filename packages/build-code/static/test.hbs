const { launch } = require("qawolf");
const selectors = require("../selectors/{{{name}}}");

describe('{{{name}}}', () => {
  let browser;

  beforeAll(async () => {
    browser = {{{launch}}}
  });

  afterAll(() => browser.close());

{{#each steps}}
  it('can {{{it}}}', async () => {
    await {{{method}}};
  });
{{#unless @last}}

{{/unless}}
{{/each}}
});