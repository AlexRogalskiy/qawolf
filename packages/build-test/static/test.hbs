describe('{{{name}}}', () => {
{{#each steps}}
  it('{{{it}}}', async () => {
    await {{{method}}};
  });
{{#unless @last}}

{{/unless}}
{{/each}}
});