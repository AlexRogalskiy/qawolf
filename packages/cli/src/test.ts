import { snakeCase } from "lodash";
import { basename } from "path";
import { runTest } from "./runTest";
import { readdir } from "fs-extra";
import { prompt } from "inquirer";

export const test = async (optionalName?: string) => {
  const testPath = `${process.cwd()}/.qawolf/tests`;
  const availableTests = (await readdir(testPath)).map(f =>
    basename(
      f
        .replace(".test", "")
        .replace(".ts", "")
        .replace(".js", "")
    )
  );

  if (!availableTests.length) {
    console.log(`No tests found in ${testPath}`);
    return;
  }

  let testName: string | null = null;

  if (optionalName) {
    testName = snakeCase(optionalName);
    if (availableTests.indexOf(testName) < 0) {
      const { selectedName } = await prompt<{ selectedName: string }>([
        {
          choices: availableTests,
          message: `Test "${testName}" not found, select a test to run`,
          name: "selectedName",
          type: "list"
        }
      ]);

      testName = selectedName;
    }
  }

  const results = await runTest(testName);
  const success = results.numFailedTestSuites < 1;
  process.exit(success ? 0 : 1);
};
