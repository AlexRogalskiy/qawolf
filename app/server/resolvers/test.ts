import { ClientError } from "../errors";
import { findLatestRuns, findRunResult } from "../models/run";
import {
  createTest,
  deleteTests,
  findTest,
  findTestForRun,
  findTestsForTeam,
  updateTest,
} from "../models/test";
import { deleteTestTriggersForTests } from "../models/test_trigger";
import {
  Context,
  CreateTestMutation,
  DeleteTestsMutation,
  TeamIdQuery,
  Test,
  TestQuery,
  TestResult,
  TestSummariesQuery,
  TestSummary,
  UpdateTestMutation,
} from "../types";
import { ensureTeamAccess, ensureTestAccess, ensureUser } from "./utils";

const ALLOW_LIST = ["flaurida", "jperl"];

/**
 * @returns The new test object
 */
export const createTestResolver = async (
  _: Record<string, unknown>,
  { team_id, url }: CreateTestMutation,
  { db, logger, user: contextUser, teams }: Context
): Promise<Test> => {
  const log = logger.prefix("createTestResolver");

  const user = ensureUser({ logger, user: contextUser });
  ensureTeamAccess({ logger, team_id, teams });

  log.debug(user.id);

  if (
    url.includes("qawolf.com") &&
    !ALLOW_LIST.includes(user.github_login || "")
  ) {
    log.error("recursion", user.id);
    throw new ClientError("recursion requires an enterprise plan");
  }

  const code = `const { context } = await launch();\nconst page = await context.newPage();\nawait page.goto('${url}', { waitUntil: "domcontentloaded" });\n// üê∫ QA Wolf will create code here`;

  return createTest(
    {
      code,
      creator_id: user.id,
      team_id: team_id,
    },
    { db, logger }
  );
};

/**
 * @returns Array of soft deleted tests
 */
export const deleteTestsResolver = async (
  _: Record<string, unknown>,
  { ids }: DeleteTestsMutation,
  { db, logger, teams }: Context
): Promise<Test[]> => {
  await Promise.all(
    ids.map((id) => ensureTestAccess({ teams, test_id: id }, { db, logger }))
  );

  return db.transaction(async (trx) => {
    await deleteTestTriggersForTests(ids, { db: trx, logger });

    return deleteTests(ids, { db: trx, logger });
  });
};

export const testResolver = async (
  _: Record<string, unknown>,
  { id, run_id }: TestQuery,
  { db, logger, teams }: Context
): Promise<TestResult> => {
  const log = logger.prefix("testResolver");

  log.debug({ id, run_id });

  if (!id && !run_id) {
    throw new Error("Must provide id or run_id");
  }

  const test = id
    ? await findTest(id, { db, logger })
    : // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      await findTestForRun(run_id!, { db, logger });

  await ensureTestAccess({ teams, test }, { db, logger });

  const runResult = run_id ? await findRunResult(run_id, { db, logger }) : null;
  if (runResult) {
    await ensureTestAccess(
      { teams, test_id: runResult.test_id },
      { db, logger }
    );
  }

  return { run: runResult, test };
};

export const testSummariesResolver = async (
  _: Record<string, unknown>,
  { test_ids, trigger_id }: TestSummariesQuery,
  { db, logger, teams }: Context
): Promise<TestSummary[]> => {
  const log = logger.prefix("testSummariesResolver");
  log.debug("tests", test_ids);

  await Promise.all(
    test_ids.map((test_id) =>
      ensureTestAccess({ teams, test_id }, { db, logger })
    )
  );

  return Promise.all(
    test_ids.map(async (test_id) => {
      const runs = await findLatestRuns(
        { test_id, trigger_id },
        { db, logger }
      );

      const lastRun = runs[0] || null;
      const gif_url = lastRun?.gif_url;

      return { gif_url, last_runs: runs, test_id };
    })
  );
};

/**
 * @returns All tests for a team, ordered alphabetically by test name ascending.
 */
export const testsResolver = async (
  _: Record<string, unknown>,
  { team_id }: TeamIdQuery,
  { db, logger, teams }: Context
): Promise<Test[]> => {
  ensureTeamAccess({ logger, team_id, teams });

  return findTestsForTeam(team_id, { db, logger });
};

export const updateTestResolver = async (
  _: Record<string, unknown>,
  args: UpdateTestMutation,
  { db, logger, teams }: Context
): Promise<Test> => {
  await ensureTestAccess({ teams, test_id: args.id }, { db, logger });
  const test = await updateTest(args, { db, logger });
  return test;
};
