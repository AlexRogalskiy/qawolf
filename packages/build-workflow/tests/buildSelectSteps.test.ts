import { loadEvents } from "@qawolf/test";
import { buildSelectSteps } from "../";

describe("buildSelectSteps", () => {
  let events: any[];

  beforeAll(async () => {
    events = await loadEvents("dropdown");
  });

  it("builds expected steps", async () => {
    const steps = buildSelectSteps(events);
    expect(steps).toMatchSnapshot();
  });
});
