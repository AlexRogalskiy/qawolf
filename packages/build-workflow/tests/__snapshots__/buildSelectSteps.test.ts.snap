// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildSelectSteps builds expected steps 1`] = `
Array [
  Object {
    "action": "select",
    "cssSelector": undefined,
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "example",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "large-12 columns",
            "id": "content",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "dropdown",
          "innertext": "Please select an option Option 1 Option 2",
        },
        "children": Array [
          Object {
            "content": "      ",
            "type": "text",
          },
          Object {
            "attrs": Object {
              "disabled": "disabled",
              "selected": "selected",
              "value": "",
            },
            "children": Array [
              Object {
                "content": "Please select an option",
                "type": "text",
              },
            ],
            "name": "option",
            "type": "tag",
            "voidElement": false,
          },
          Object {
            "attrs": Object {
              "value": "1",
            },
            "children": Array [
              Object {
                "content": "Option 1",
                "type": "text",
              },
            ],
            "name": "option",
            "type": "tag",
            "voidElement": false,
          },
          Object {
            "attrs": Object {
              "value": "2",
            },
            "children": Array [
              Object {
                "content": "Option 2",
                "type": "text",
              },
            ],
            "name": "option",
            "type": "tag",
            "voidElement": false,
          },
        ],
        "name": "select",
        "type": "tag",
        "voidElement": false,
      },
    },
    "index": 1,
    "page": 0,
    "value": "2",
  },
]
`;
